coffee/cse.py

```diff
@@ -0,0 +1,419 @@
+        finder = FindInstances(Symbol)
+        syms = finder.visit(expr, ret=FindInstances.default_retval())[Symbol]
+        syms = [s for s in syms
+                if any(l in self.expr_info.linear_dims for l in lda[s])]
+
+        syms_costs = defaultdict(int)
+
+        def wrapper(node, found=0):
+            if isinstance(node, Symbol):
+                if node in syms:
+                    syms_costs[node] += found
+                return
+            elif isinstance(node, (EmptyStatement, ArrayInit)):
+                return
+            elif isinstance(node, (Prod, Div)):
+                found += 1
+            operands = zip(*explore_operator(node))[0]
+            for o in operands:
+                wrapper(o, found)
+        wrapper(expr)
+
+        return syms_costs
+
+    def _analyze_loop(self, loop, nest, lda, global_trace):
+        trace = OrderedDict()
+
+        for node in loop.body:
+            if not isinstance(node, Writer):
+                not_ssa = [trace[w] for w in in_written(node, key='urepr') if w in trace]
+                for t in not_ssa:
+                    t.is_read.append(t.symbol)
+                continue
+            syms_costs = self._analyze_expr(node.rvalue, lda)
+            for s in syms_costs.keys():
+                if s.urepr in global_trace:
+                    temporary = global_trace[s.urepr]
+                    temporary.is_read.append(node.lvalue)
+                    temporary = temporary.reconstruct()
+                    temporary.level = -1
+                    trace[s.urepr] = temporary
+                else:
+                    temporary = trace.setdefault(s.urepr, Temporary(s, loop, nest))
+                    temporary.is_read.append(node.lvalue)
+            new_temporary = Temporary(node, loop, nest, syms_costs)
+            new_temporary.level = max([trace[s.urepr].level for s
+                                       in new_temporary.reads] or [-2]) + 1
+            trace[node.lvalue.urepr] = new_temporary
+
+        return trace
+
+    def _group_by_level(self, trace):
+        levels = defaultdict(list)
+
+        for temporary in trace.values():
+            levels[temporary.level].append(temporary)
+        return levels
+
+    def _cost_cse(self, levels, bounds=None):
+        if bounds is not None:
+            lb, up = bounds[0], bounds[1] + 1
+            levels = {i: levels[i] for i in range(lb, up)}
+        cost = 0
+        for level, temporaries in levels.items():
+            cost += sum(t.flops*t.niters for t in temporaries)
+        return cost
+
+    def _cost_fact(self, trace, levels, bounds):
+        # Check parameters
+        bounds = bounds or (min(levels.keys()), max(levels.keys()))
+        assert len(bounds) == 2 and bounds[1] >= bounds[0]
+        assert [i in levels.keys() for i in bounds]
+        fact_levels = OrderedDict([(k, v) for k, v in levels.items()
+                                   if k > bounds[0] and k <= bounds[1]])
```

Comment:
It's unclear why you would want an ordered dict here given that levels is an unordered dict.